mixer/adapter/svcctrl/svcctrl.go

```diff
@@ -12,115 +12,170 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package svcctrl // import "istio.io/istio/mixer/adapter/svcctrl"
+package svcctrl
 
 import (
-	"bytes"
 	"context"
-	"time"
+	"errors"
+	"fmt"
 
-	"github.com/pborman/uuid"
-	sc "google.golang.org/api/servicecontrol/v1"
+	pbtypes "github.com/gogo/protobuf/types"
+	multierror "github.com/hashicorp/go-multierror"
 
 	"istio.io/istio/mixer/adapter/svcctrl/config"
+	"istio.io/istio/mixer/adapter/svcctrl/template/svcctrlreport"
 	"istio.io/istio/mixer/pkg/adapter"
-	"istio.io/istio/mixer/template/metric"
+	"istio.io/istio/mixer/template/apikey"
+	"istio.io/istio/mixer/template/quota"
 )
 
-type handler struct {
-	serviceControlClient *sc.Service
-	env                  adapter.Env
-	configParams         *config.Params
+// svcctrl adapter builder
+type builder struct {
+	config          *config.Params // Handler config
+	checkDataShape  map[string]*apikey.Type
+	reportDataShape map[string]*svcctrlreport.Type
+	quotaDataShape  map[string]*quota.Type
 }
 
```

Comment:
consider adding compile time checkers here
```
var _ apikey.ApiKeyBuilder = builder{}
```

same for the handler implementation.